### PR TITLE
dev-refactor merge 

### PR DESCRIPTION
**Pull request by felsweg-iota**
# Description of change

This is the branch for upcoming features. No explicit change, but to keep `dev-refactor` mergable with `dev`. Merge this PR into `dev`, when all other changes are completed. 

## Type of change

- [x] Enhancement (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Fix

## How the change has been tested

All dependent branches and features have thoroughly been tested by unit and integration tests. 

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .github/workflows/covector-version-or-publish.yml
Comment (elenaf9) [inline .github/workflows/covector-version-or-publish.yml:13]: Why do we need the covector workflow for dev-refactor?

Comment (felsweg-iota) [reply .github/workflows/covector-version-or-publish.yml:13]: Agreed. That shouldn't be part of `dev-refactor`.

**Review by vuongDang — APPROVED** [review]

File: engine/new_runtime/src/memories/mod.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/mod.rs]: Nit: Move the mod file from `src/memories/mod.rs` -> `src/memories.rs` to be consistent in the module file naming throughout the crate.

File: engine/new_runtime/src/memories/buffer.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/buffer.rs]:
```suggestion
/// [`Buffer`] is used when we want to store sensitive non encrypted data.
/// This shall always be short lived.
#[derive(Clone, Eq)]
pub struct Buffer<T: Bytes> {
    boxed: Boxed<T>, // the boxed type of current Buffer
}
```

Comment (elenaf9) [reply engine/new_runtime/src/memories/buffer.rs]: Also, maybe re-add the old comment about the security features and guarantees?: 
```suggestion
/// [`Buffer`] is used when we want to store sensitive non encrypted data.
/// This shall always be short lived.
///
/// Provides the following features and guarantees:
/// * Causes segfault upon access without using a borrow.
/// * Protected using mprotect:
///    * `Prot::NoAccess` - when the box has no current borrows.
///    * `Prot::ReadOnly` - when the box has at least one current immutable borrow.
///    * `Prot::ReadWrite` - when the box has a current mutable borrow (can only have one at a time).
/// * The allocated memory uses guard pages both proceeding and following the memory. Overflows and large underflows
///   cause immediate termination of the program.
/// * A canary proceeds the memory location to detect smaller underflows.  The program will drop the underlying memory
///   and terminate if detected.
/// * The Memory is locked with `mlock`.
/// * When the memory is freed, `munlock` is called.
/// * The memory is zeroed when no longer in use via `sodium_free`.
/// * `Buffer` types can be compared in constant time.
/// * `Buffer` types can not be printed using `Debug`.
/// `Buffer` includes serialization which converts the data into a vector before its serialized by serde.  Upon
/// deserialization, the data is returned back to a new `Buffer`.
#[derive(Clone, Eq)]
pub struct Buffer<T: Bytes> {
    boxed: Boxed<T>, // the boxed type of current Buffer
}
```

File: engine/new_runtime/src/memories/buffer.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/buffer.rs]: Since `Buffer` is public I don't think we need  `#[allow(dead_code)]` for any of its methods.

File: engine/new_runtime/Cargo.toml
Comment (elenaf9) [inline engine/new_runtime/Cargo.toml]: A `.license_template` is missing for `new_runtime`.

File: engine/new_runtime/Cargo.toml
Comment (elenaf9) [inline engine/new_runtime/Cargo.toml]:
```suggestion
serde_json = "1.0"

```
Nit: newline at end of file

File: engine/new_runtime/src/memories/ram_memory.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/ram_memory.rs]:
```suggestion
/// Protected ram memory that may be encrypted or not.
/// This is basically a wrapper for the [`Buffer`] type, but the usage
/// is different. The `Buffer` type is meant for short lived usage while
/// `RamMemory` can store data for longer period of time.
/// Hence data in `RamMemory` has to be either encrypted or protected
/// behind a scheme.
```

File: engine/new_runtime/src/memories/ram_memory.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/ram_memory.rs]:
```suggestion
    /// Locks the memory and possibly reallocates.
    // Currently we reallocate a new [`RamMemory`] at each lock.
    // This improves security but decreases performance.
```

File: engine/new_runtime/src/memories/ram_memory.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/ram_memory.rs]:
```suggestion
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RamMemory {
```

No need to manually implement `Debug`, `Serialize` and `Deserialize` below. They are already implemented on `Buffer`, so we can just derive them here.

File: engine/new_runtime/src/memories/ram_memory.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/ram_memory.rs]:
```suggestion
```

`Send` and `Sync` are automatically derived since `Buffer` already implements them. Quoting the Rustonomicon:

> Send and Sync are also automatically derived traits. This means that, unlike every other trait, if a type is composed entirely of Send or Sync types, then it is Send or Sync. Almost all primitives are Send and Sync, and as a consequence pretty much all types you'll ever interact with are Send and Sync.

File: engine/new_runtime/src/memories/file_memory.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/file_memory.rs]: ...and what? :D

File: engine/new_runtime/src/memories/noncontiguous_memory.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/noncontiguous_memory.rs]: I don't think we need the `config` property here, it is only used on `alloc`. And if we need it again, we can just derive it by matching on the variants of `shard1` and `shard2`.

File: engine/new_runtime/src/memories/noncontiguous_memory.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/noncontiguous_memory.rs]:
```suggestion
    pub fn refresh(self) -> Result<Self, MemoryError> {
```

Shouldn't this method be public to the user?

File: engine/new_runtime/src/memories/noncontiguous_memory.rs
Comment (elenaf9) [inline engine/new_runtime/src/memories/noncontiguous_memory.rs]: Amazing work with the non-contiguous memory!

File: engine/src/vault/view.rs
Comment (elenaf9) [inline engine/src/vault/view.rs:519]:
```suggestion
```

Remove old code.

File: engine/src/vault/crypto_box.rs
Comment (elenaf9) [inline engine/src/vault/crypto_box.rs:84]: Why is this method removed? I think we need such a convenience method again, to avoid that the user has to the awkward call of:
```
let key = *(&key.key).borrow()
```
to access the actual key.

File: engine/src/vault/crypto_box.rs
Comment (elenaf9) [inline engine/src/vault/crypto_box.rs:49]: I know in practice it does not make a difference, but should we use the `RamMemory` type here because the `Key` is stored for a longer period of time? I thought the `Buffer` type should only be used when accessing a value in a `LockedMemory`.

File: engine/src/vault/crypto_box.rs
Comment (elenaf9) [inline engine/src/vault/crypto_box.rs]: Shouldn't `engine/lib.rs` export this type? In which case we also don't need `#[allow(dead_code)]` on any of its public mehtods.

File: engine/src/vault/key_store.rs
Comment (elenaf9) [inline engine/src/vault/key_store.rs]: Nice, this also solves issue #268!

File: engine/src/vault/key_store.rs
Comment (elenaf9) [inline engine/src/vault/key_store.rs]: Again, I think `KeyStore` should also be public outside of the crate so we can use it in `iota_stronghold`. In that case we also don't need the `#![allow(dead_code)]`.

**Review by elenaf9 — APPROVED** [review]
Really great work!

Sorry for my late review here, did not have time to look through #316 before it was merged.
Couple of comments. Almost all are nitpicks, the only blocking ones are that `engine::vault::{KeyStore, NCKey}` should be public if we want to use them in the client.

File: engine/src/vault/key_store.rs
Comment (elenaf9) [reply engine/src/vault/key_store.rs]: Resolved with #324.

File: engine/src/vault/crypto_box.rs
Comment (elenaf9) [reply engine/src/vault/crypto_box.rs]: Resolved with #324.

File: engine/new_runtime/src/memories/file_memory.rs
Comment (tensor-programming) [reply engine/new_runtime/src/memories/file_memory.rs]: haha

File: client_new/src/types/stronghold/base.rs
Comment (tensor-programming) [inline client_new/src/types/stronghold/base.rs]: I know these aren't the complete interfaces for the user to interact with, but maybe it would be a little bit more clear how a user should interact with this library if the stronghold type and interfaces weren't nested so far in a folder like this.  The same for the other front facing types, client, keystore etc.  Or at least some docs that highlight that they are the main types that should be used.

File: client_new/src/tests/interface_tests.rs
Comment (tensor-programming) [inline client_new/src/tests/interface_tests.rs]: I wonder, do these tests even need to be a part of the core program anymore or can they be put into a test file?  The reason before was because there were some items things that weren't public which needed to be tested through the interface.

Comment (felsweg-iota) [reply client_new/src/tests/interface_tests.rs]: could be moved into an integration tests, yes

File: client_new/src/types/stronghold/base.rs
Comment (felsweg-iota) [reply client_new/src/types/stronghold/base.rs]: yeah, docs coming up. i mean for the `Stronghold` type the intention was to separate `base` impl. for all local  and `p2p` for all (upcoming / ported) networking functionality.

File: engine/new_runtime/src/memories/file_memory.rs
Comment (felsweg-iota) [reply engine/new_runtime/src/memories/file_memory.rs]: you're so mean :rofl:

File: client_new/src/types/client.rs
Comment (elenaf9) [inline client_new/src/types/client.rs]: With #319 all operations on the vault can be done through procedures. We could consider removing `ClientVault` so that we have a single way of doing things through procedures only.

File: client_new/src/types/client.rs
Comment (elenaf9) [inline client_new/src/types/client.rs]: Why are some of the methods on the `Client` async? Unless I am missing something, there is no need for it.

File: client_new/src/types/client.rs
Comment (felsweg-iota) [reply client_new/src/types/client.rs]: `ClientVault` is a thin wrapper around `Client` and calls all actions directly on `Client`s database.

Comment (elenaf9) [reply client_new/src/types/client.rs]: Yes but said actions (`write_secret`, `delete_secret`, `revoke_secret` and `cleanup`) can all be done through procedures (`StrongholdProcedure::WriteVault`, `StrongholdProcedure::RevokeData`, etc). 
Thus the methods are kinda redundant and if we would remove them, we also would not need this `ClientVault` wrapper anymore.

Comment (felsweg-iota) [reply client_new/src/types/client.rs]: I disagree. The `ClientVault` is meant to have as a distinct type to differentiate the actions taken on them. The whole idea of the new client interface is around types.

Comment (felsweg-iota) [reply client_new/src/types/client.rs]: The only point to argue would be to expose this `ClientVault` as an interface on `Client`.

File: client_new/src/types/client.rs
Comment (felsweg-iota) [reply client_new/src/types/client.rs]: true, it's a reminiscent of the old interface. Most of the operations are sync, since we do not enforce a hard dependency on some async runtime. I/O operations on the `Snapshot` should be async, imho. I wonder, if we should enforce a dependency on tokio (or any other async runtime)

File: client_new/src/types/client.rs
Comment (elenaf9) [reply client_new/src/types/client.rs]: My argument was based on what is described in the _Primitive Based Interface_ RFC in #313 in **Reference-level explanation - Vault**: 

> While the vault is a stronghold primitive, it does not have its own type (being the only exception). Most interesting operations on a vault are procedures. Since a client can have multiple vaults, a procedure generally works on the client-level rather than the vault-level (...)

I think it's cleaner to have a unified way of interacting with vaults by always using procedures. It's also easier for the user to understand, rather than having multiple ways of how something can be written to a vault.

But I also see your point. Having a `ClientVault` type is more consistent with how the interface is generally designed.

Anyway, your call 😄.

Comment (felsweg-iota) [reply client_new/src/types/client.rs]: The execution of procedures are on `Client` level, while `ClientVault` offers a more distinct type :D Sure, every action can be done by procedures, but having access to "simpler" function calls might be beneficial, imho.

File: engine/src/vault/crypto_box.rs
Comment (vuongDang) [reply engine/src/vault/crypto_box.rs:49]: Yes you're right, will change it

**Pull request merged**
